### PR TITLE
Add SmallVector class.

### PR DESCRIPTION
**Pull request by mzient**
Add a `SmallVector` class.
* compile-time static storage
* can grow beyond static storage
* support insert, erase, push/pop, subscript
* usable from host and device code (with appropriate alloator)
* capacity stored in static storage to reduce size.

### CODE REVIEW

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: include/dali/core/small_vector.h [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+class SmallVectorBase<T, true> {
+ protected:
+  __host__ __device__ void move_and_destroy(T *dst, T *src, size_t count) noexcept
+  {
+#ifdef __CUDA_ARCH__
+    for (size_t i = 0; i < count; i++) {
+      dst[i] = src[i];
+    }
+#else
+    memcpy(dst, src, count * sizeof(T));
```
Comment (klecki) [inline include/dali/core/small_vector.h]: Isn't there a device side memcpy as well? Is it slow or problematic to dispatch here?

Comment (mzient) [reply include/dali/core/small_vector.h]: It's slow, according to multiple posts on stackoverflow.

**Comment by mzient** [comment]
!build

File: include/dali/core/small_vector.h [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+      if (is_dynamic() && v.size() <= static_size) {
+        deallocate(dynamic_data(), capacity());
+        reset_capacity();
+      } else {
+        reserve(v.size());
+      }
+      T *src = v.data();
+      T *dst = data();
+      if (std::is_pod<T>::value) {
+
```
Comment (jantonguirao) [inline include/dali/core/small_vector.h]: ??

Comment (mzient) [reply include/dali/core/small_vector.h]: Apparently, not implemented (!)

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Pull request merged**